dvc/remote/base.py

```diff
@@ -547,11 +547,51 @@ def download(
         if to_info.scheme != "local":
             raise NotImplementedError
 
-        logger.debug("Downloading '{}' to '{}'".format(from_info, to_info))
+        if self.isdir(from_info):
+            return self._download_dir(
+                from_info, to_info, name, no_progress_bar, file_mode, dir_mode
+            )
+        return self._download_file(
+            from_info, to_info, name, no_progress_bar, file_mode, dir_mode
+        )
 
-        name = name or to_info.name
+    def _download_dir(
+        self, from_info, to_info, name, no_progress_bar, file_mode, dir_mode
+    ):
+        file_to_infos = (
+            to_info / file_to_info.relative_to(from_info)
+            for file_to_info in self.walk_files(from_info)
+        )
 
+        with ThreadPoolExecutor(max_workers=self.JOBS) as executor:
+            file_from_info = list(self.walk_files(from_info))
```

Comment:
Hate to resurrect this, but walking files twice is unsafe, if some file will be added or removed in between then file will be written under their neighbors names. This might be a spectacular bug ;)

Also, this is inefficient.

Reply:
@Suor Good point! It was pointed out earlier in https://github.com/iterative/dvc/pull/2894#discussion_r354507715 , but looks like got lost during the iterations. I'll do a quickfix soon.

Reply:
@Suor Created https://github.com/iterative/dvc/pull/2909 .